File: Arrays/03 Sorted Squared Array/my_solution.js
// Qns :- Given an array on numbers, square each number and sort the result in ascending order. in array format

// sol 1 O(nlogn) time | O(n) space
// ex 1

// const sortedSquaredArray = (arr) => {
//   return arr.map((num) => num * num).sort((a, b) => a - b);
// };

// ex2

// const sortedSquaredArray = (arr) => {
//   arr.sort((a, b) => a - b);
//   let sortedSquaredArray = arr.map(() => 0);
//   for (let i = 0; i < arr.length; i++) {
//     let value = arr[i];
//     sortedSquaredArray[i] = value * value;
//   }
//   sortedSquaredArray.sort((a, b) => a - b);
//   return sortedSquaredArray;
// };

// sol 2 O(n) time | O(n) space

// const sortedSquaredArray = (arr) => {
//   let squared = arr.map(() => 0);
//   let left = 0;
//   let right = arr.length - 1;
//   let index = arr.length - 1;
//   while (left <= right) {
//     let left_square = arr[left] * arr[left];
//     let right_square = arr[right] * arr[right];
//     if (left_square > right_square) {
//       squared[index] = left_square;
//       left += 1;
//     } else {
//       squared[index] = right_square;
//       right -= 1;
//     }
//     index -= 1;
//   }

//   return squared;
// };

const sortedSquaredArray = (arr) => {
  let squared = arr.map(() => 0); // [0,0,0,0,0,0,0,0,0,0]
  let smallerValueIndex = 0; // assigning pointer to start of the array
  let largerValueIndex = arr.length - 1; // assigning pointer to end of the array

  for (let i = arr.length - 1; i >= 0; i--) {
    // iterating backwards on Given array
    let smallerValue = arr[smallerValueIndex]; //getting the value of the pointer [first value in the array]
    let largerValue = arr[largerValueIndex]; //getting the value of the pointer [last value in the array]
    if (Math.abs(smallerValue) > Math.abs(largerValue)) {
      //checking if the absolute value of the smaller value is greater than the absolute value of the larger value
      squared[i] = smallerValue * smallerValue; // if true, then multiply the smaller value by itself and assign it to the current index of the squared array
      smallerValueIndex += 1; // increment the pointer
    } else {
      // if the smaller value is greater than the larger value
      squared[i] = largerValue * largerValue; //then multiply the larger value by itself and assign it to the current index of the squared array
      largerValueIndex -= 1; //increment the pointer
    }
  }
  return squared; //return the squared array
};

console.log(sortedSquaredArray([-4, -1, 0, 3, 10]));

File: Arrays/03 Sorted Squared Array/my_solution.py
# sol 1 O(nlogn) time | O(n) space
# def sortedSquaredArray(A):
#     return sorted([i*i for i in A])

# def sortedSquaredArray(A):
#     A.sort()
#     sortedSquares = [0 for _ in A]

#     for idx in range(0, len(A)):
#         value = A[idx]
#         sortedSquares[idx] = value * value
#     sortedSquares.sort()
#     return sortedSquares

# sol 2 O(n) time | O(n) space
# def sortedSquaredArray(array):
#     squared = [0 for _ in array]
#     left = 0
#     right = len(array) - 1
#     idx = len(array) - 1
#     while left <= right:
#         left_square = array[left] * array[left]
#         right_square = array[right] * array[right]
#         if left_square > right_square:
#             squared[idx] = left_square
#             left += 1
#         else:
#             squared[idx] = right_square
#             right -= 1
#         idx -= 1
#     return squared

def sortedSquaredArray(array):
    sortedSquares = [0 for _ in array]
    smallerValueIdx = 0
    largerValueIdx = len(array) - 1

    for idx in reversed(range(len(array))):
        smallerValue = array[smallerValueIdx]
        largerValue = array[largerValueIdx]

        if abs(smallerValue) > abs(largerValue):
            sortedSquares[idx] = smallerValue * smallerValue
            smallerValueIdx += 1
        else:
            sortedSquares[idx] = largerValue * largerValue
            largerValueIdx -= 1
    return sortedSquares


print(sortedSquaredArray([-4, -1, 0, 3, 10]))
